WebSocket support for Carbon-Messaging

irunika opened the pull request:
The current implementation of carbon-messaging does not support WebSocket messaging.
We need to add WebSocket support to carbon-messaging in order to decouple [Carbon-Transport](https://github.com/wso2/carbon-transports) from [WSO2 MSF4J](https://github.com/wso2/msf4j).

So in order to get support,
New Carbon Message types are added to Carbon-Messaging. These can be used in any protocol which matches to the content of the message.

- **TextCarbonMessage**

- **BinaryCarbonMessage**

- **StatusCarbonMessage**

- **ControlCarbonMessage**

This is increasing the carbon-messaging to next major version since new message types are added. 

resolves #38

irunika committed:
Added :
WebSocketCarbonMessage.java
TextWebSocketCarbonMessage.java
BinaryWebSocketCarbonMessage.java
CloseWebSocketCarbonMessage.java
WebSocketResponder.java

These will be some of the Carbon Messaging components of handling WebSocket

irunika committed:
Added minor changes to comments

irunika committed:
Added new classes :  PingWebSocketCarbonMessage.java, PongWebsocketCarbonMessage.java to handle WebSocket ping and pong messages

irunika committed:
Minor change to PongWebsocketCarbonMessage.java

irunika committed:
Added minor changes to comments in PingWebSocketCarbonMessage.java and PongWebsocketCarbonMessage.java

irunika committed:
Changed naming convention from WebSocketCarbonMessage to WebSocketMessage. Added (.)dot to the ends of comments. Added new doc comments  and expanded some comments to make the implementation clear.

irunika committed:
Modified the comments of BinaryWebSocketMessage.java and CloseWebSocketMessage.java . Added new experimental message type WebSocketHandshakeMessage.java to transfer handshaking from transport layer to MSF4J

irunika committed:
Working modal of Carbon-Messaging in MSF4J with WebSocket. Cannot have transport oriented data. They have to be changed.

irunika committed:
Changed the naming conventions. Removed protocol related stuff since Messaging does not depend on a protocol.

Kishanthan commented on components/src/main/java/org/wso2/carbon/messaging/BinaryCarbonMessage.java:
```diff
@@ -0,0 +1,55 @@
+import java.nio.ByteBuffer;
+
+/**
+ * {@link CarbonMessage} type for Binary Message.
```
Need a new line in between

irunika replied:
Done.

Kishanthan commented on components/src/main/java/org/wso2/carbon/messaging/BinaryCarbonMessage.java:
```diff
@@ -0,0 +1,55 @@
+/*
+ *   Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
All the lines in the license header seems to be not aligned properly

irunika committed:
Did corrections of issues in License header and doc comments.

Kishanthan committed:
adding/improving trasport listener APIs to handle polling type scenarios

Kishanthan committed:
improving transport listener APIs to handle errors

Kishanthan committed:
changing license header year value

Kishanthan committed:
adding TransportListenerManager, removing dependency on kernel for TransportAPIs

Kishanthan committed:
adding setter for id a and default constructor

irunika committed:
Replaced Ping and Pong Carbon Messages from ControlCarbonMessage since a general control message can be used for both of them.

isudana commented on components/src/main/java/org/wso2/carbon/messaging/CloseCarbonMessage.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * {@link CarbonMessage} type for Close Messages.
+ */
+public class CloseCarbonMessage extends CarbonMessage {
```
Can we make this generic ?  `CloseCarbonMessage` sounds more web sockets specific one. 
Can we use Control message or something similar for this ?

irunika replied:
@isudana 
We have used ControlCarbonMessage for Ping pong already. Are you suggesting to use the same CarbonMessageType for CloseCarbonMessage?
The problem is it will include status code and close reason. 
We can add them as properties. But there will be an overhead on the application side to check all those things.
WDYT?

isudana replied:
The problem with CloseCarbonMessage is, it is not going inline with rest of the Message Types (Binary, Text, etc.). 
Renaming it to a more generic one or enclosing it with a more generic one (like what you did for Ping and Pong i.e. Ping + Pong --> ControlMessage) is what I am suggesting.

irunika replied:
But the most generic name I can think of is CloseMessage / CloseCarbonMessage.
It is a carbon message when someone closing the connection 
It can be used in any protocol which provides a status code and reason text (Not necessary).
If we go with control message then it will be a performance overhead for application since it would have to check whether the message is a _ping, pong_ or a _close_.
Then if it is close, then it will have to get the close status code and reason text as properties, cast them and use them.
With the current approach, it is only checking the instance of it and if it is a close message it simply call the _getStatusCode()_ and _getReasonText()_.

isudana replied:
Can you please point me to a usage of this ?

irunika replied:
@isudana 
As you suggested in the offline chat it is better to use a general message type for CloseMessage too. So as suggested I will create new message type StatusCarbonMessage which we can use for both open and close status for now.

irunika replied:
Hi all,
I did the necessary changed. Please review and merge.
Thanks

ramith commented on components/src/main/java/org/wso2/carbon/messaging/BinaryCarbonMessage.java:
```diff
@@ -0,0 +1,54 @@
+ */
+public class BinaryCarbonMessage extends CarbonMessage {
+
+    private ByteBuffer bytes; //ByteBuffer to store binary data
```
why not properly document these ?

Kishanthan committed:
removing no-arg constructor

Kishanthan committed:
removing getProtocol and setId methods as they are no longer used by any listener

irunika committed:
Merge remote-tracking branch 'upstream/server-connector-fw-improvements' into websocket-support

irunika committed:
\Changed CloseCarbonMessage to StatusCarbonMessage which can be used to send any status of a connection. This makes huge performance improvement in app level too.
Added inputStream support for TextCarbonMessage
Added necessary constants for StatusCarbonMessage

irunika committed:
Modified the getInputStream() in TextCarbonMessage

irunika committed:
Modified the doc comments of the Message types.

isudana commented on components/src/main/java/org/wso2/carbon/messaging/TextCarbonMessage.java:
```diff
@@ -0,0 +1,60 @@
+import java.nio.charset.StandardCharsets;
+
+/**
+ * {@link CarbonMessage} type for WebSocket Text Messages. This message type is better if you are dealing with
```
Need to remove word "WebSocket" here in javadoc comment

irunika replied:
Thanks. 
Done.

irunika committed:
Monor change in BinaryCarbonMessage.java

irunika committed:
Monor change for  the java docs of TextCarbonMessage.java
